msyslog-v1.08a+smac/src/man/im_file.8
.\"	$CoreSDI: im_file.8,v 1.1.2.1 2001/11/21 06:37:31 alejo Exp $
.\"
.\" Copyright (c) 2000, 2001
.\"	Core-SDI SA. All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. Neither the name of Core-SDI SA nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.Dd Nov 06, 2001
.Dt IM_FILE 8
.Os Core-SDI
.Sh NAME
.Nm FILE input module
.Nd
.Xr syslogd 8
input module for reading files and pipes
.Sh SYNOPSIS
.Nm file
.Op Fl f Ar path
.Op Fl p Ar path
.Op Fl n Ar program
.Sh DESCRIPTION
Reads the log messages from a file/pipe writen by another program
for processing by the
.Xr syslogd 8
daemon. It's options are:
.Bl -tag -width Ds
.It Ar log_socket
Socket to use, default is /dev/log.
.Sh EXAMPLES
.Bl -bullet
.It
The following command installs
.Xr syslogd 8
daemon reading from file /some/path:
.Pp
.Dl syslogd -i 'file -f /some/path'
.El
.Bl -bullet
.It
Same as above, but sets the name to 'myapp':
.Pp
.Dl syslogd -i 'file -n myapp -f /some/path'
.El
.Bl -bullet
.It
Same as above, but on /some/named/pipe pipe instead of the file:
.Pp
.Dl syslogd -i 'file -n myapp -p /some/named/pipe'
.El
.Sh SEE ALSO
.Xr syslog 3 ,
.Xr im_bsd 8 ,
.Xr im_doors 8 ,
.Xr im_linux 8 ,
.Xr im_streams 8 ,
.Xr im_tcp 8 ,
.Xr im_udp 8 ,
.Xr syslogd 8
.Sh BUGS
.Bl -bullet
.It
Submit bugs at this project's Sourceforge Bug reporting system at:
http://sourceforge.net/tracker/?func=add&group_id=25741&atid=385117
You may also report them directly to the authors; send an email to
core.devel.alat@corest.com, describing the problem the most you can,
containing also machine description, hardware description, the
configuration file (/etc/syslog.conf), the OS description, and the
invoking command line.
The more you describe the bug, the faster we can fix it.
.El


html/ChangeLog.txt
$Id$

Feature Request:

Bugs:
1.  Change code to allow for tailing spaces when looking at the expression field

3/10/2004
	-  Removing any reference to company name

3/9/2004
	-  Changed out company logo
	-  Added GPL notice to all libraries

1/21/2004
	-  Found a bug where a host would email off an alert with no text.  I've put a stop gap fix in.

1/13/2004
        -  Fixed bug with msyslog.pgsql where SET STORAGE didn't have a space before it.
        -  Fixed bug where hour in time of rules would default to 18:00 for no good reason
	-  Fixed two problems with cloning rules:  1)  order was not preserved  2)  an imported version of the database couldn't clone because fields that were empty needed to have non-null defaults applied
	-  Forgot to merge in latest VPN reports

1/6/2004
        -  Added 'left menu' support for ACID and MRTG groups in security framework should the ever be added 8)

1/6/2004
	-  Added 'left menu' support for ACID and MRTG groups in security framework should the ever be added 8)

1/1/2004
	-  Fixed spelling error in with the word 'threshold' in the rules schema
	-  Fixed a bug in vacuumdb where it was 'ANALYZ' not 'ANALYZE'

12/19/2003
	-  Updated code to look into /opt/apache

12/10/2003
	-  Updated database to schema to not use compressed text fields(we'll see how this performs)
	-  working on adding interfaces to more lock data and other new stats with PostgreSQL V7.4
	-  cleaned up some button descriptions on the maintenance page, also added a lock view as well as a settings view

12/9/2003
	-  Finished rule.php support for basic timer maintenance.  Need to add another page to graft timers onto rules.
	-  processlogs.php is now setup to support rule timers
	-  need to write rule expiration process 
	-  started converting away from compressed text in hopes of providing faster data retrieval.... also pulled OIDs from DB definition

12/8/2003
	-  Adding support for date and time based rules with date ranges, day of week selections, as well as deleted rules, need to add interface to control rule timestamp properties

12/7/2003
	-  finished adding accumulation thresholds

12/3/2003
	-  adding support for both types of thresholds.  Need to update web pages to reflect new radio buttons.
	-  added web configuration support for supression thresholds and accumulating thresholds, now onto updating the log processor
	-  basic supression works(kinda), need to verify functionality

12/2/2003
	-  customer profiles can now have multiple hosts added at once.
	-  fixed a stupid bug where <prev|refresh|next> didn't behave right.  next would stop working if you started at oct-26-2003 and it would stay on oct-26-2003
	-  Added web-based framework and database schema to support alert supression thresholds

12/1/2003
	-  Updated processlogs.php code to better deal with single entry/no rules vs mutliple rules

11/19/2003
	-  Updated mail table to enforce unique login ids(effectively stopping two processes from running at the same time, one will crash and die(safely))
	-  Updated openmail and closeopenmail to use transaction support since PostgreSQL no longer does server-side auto-commit(ie. convert everything to transactions)
	-  Transaction support should now be officially added, will do some testing

11/18/2003
	-  Updating program for support with PostGreSQL V7.4
	-  Fixed host process table to reflect the fact that multiple hosts are in there by default	

10/13/2003
	-  Added alert total to bottom of alert aggregation

10/6/2003
	-  finished support for alert aggregation
	-  updated 1stview to pull the current time and date
	-  fixed host.php bug where you could expire syslogs but not be forced to expire syslogs

10/5/2003
	-  still adding support for alert aggregation.  Basic aggregation works plus alert zooming but need to add support for across the board for other alert queries

10/3/2003
	-  added ability to un/suspend log processors from the web interface
	-  updated maintenance to rebuild all indexes in an better manner(ie. grab the index list from the DB rather then by hand)
	-  added additional framework to do alert aggregation interface.  Need to add 'aggregation code' for display

9/24/2003
	-  weeklyindexrebuild.php now pulls all indexes from the system and rebuilds them.  The result is that the system will now rebuild any new indexes without manual reconfiguration.

9/15/2003
	-  launchid was not initiated correctly in the clonerule.

9/2/2003
	-  Fixed a bug where '\'s at the end of a line caused problems because we were not properly dealing with them in general.  Fixed that.

8/23/2003	
	-  Created another bug when fixing 5000 line paging.  Timestamp was thrown off in view.php

8/18/2003	
	-  Syncing changes from production smt environment:  vacuumtsyslog.php
	-  Updated weeklyindexrebuild.php to account for the correct indexes
	-  Updated maint.php to account for the three new indexes for the launch program section

8/13/2003
	-  view.php has had several updates.  Paging should now be fixed.  Multiple searches appeared not to be working correctly.
	-  Needed to add lastid as hidden var if the variable was set
	-  Needed to use urlencode on top of htmlspecialcharacters, filters were broken because of it

8/11/2003
	-  1stfilter.php doesn't list 'global' filters that you down own
	-  modified view.php to not let the user save a filter with no description
	-  added support to delete all of a user's filters(ie. do before delete)
	-  another problem popped up with filter.php when I added the delete user filter option

8/1/2003	
	-  processlogs.php now supports launching external programs!

7/31/2003
	-  Updated vacuumdb.php to do a full vacuum of the TSyslog table.  Why?  Because the system doesn't reclaim disk space or use old delete space for some reason
	-  Almost finished adding launch program code, need to test.

7/27/2003
	-  Added weeklyindexrebuild.php which rebuilds all indexes at 5am Sunday morning

7/23/2003

msyslog-v1.08a+smac/src/modules/im_file.c
/*	$CoreSDI: im_file.c,v 1.11.2.1 2001/11/21 06:34:27 alejo Exp $	*/

/*
 * Copyright (c) 2001, Core SDI S.A., Argentina
 * All rights reserved
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 *
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. Neither name of the Core SDI S.A. nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
 * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
 * OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
 * IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT,
 * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
 * NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
 * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
 * THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

/*
 * im_file -- read from a log file being written by other program
 *
 * Author: Alejo Sanchez for Core-SDI SA
 * http://www.corest.com/
 *
 */

#include "config.h"

#include <sys/types.h>
#include <ctype.h>
#include <errno.h>
#include <fcntl.h>
#include <limits.h>
#include <signal.h>
#include <stdlib.h>
#include <string.h>
#include <syslog.h>
#include <time.h>
#include <unistd.h>

#include "../modules.h"
#include "../syslogd.h"

struct im_file_ctx {
	char	*timefmt;
	int	start;
};

/*
 * initialize file input
 *
 */

int
im_file_init(struct i_module *I, char **argv, int argc)
{
	char			*path;
	struct im_file_ctx	*c;
	int			ch, optind_s, start;

	dprintf(MSYSLOG_INFORMATIVE, "im_file_init: Entering\n");

	I->im_path = NULL;
	start = 0;
	path = NULL;
	c = NULL;	/* just to make compiler happy */

	/* parse command line */
	optind_s = optind;	/* save main's optind */
	optind = 1;
#ifdef HAVE_OPTRESET
	optreset = 1;
#endif
	while ((ch = getopt(argc, argv, "f:p:n:t:s:")) != -1) {
		switch (ch) {
		case 'f':
			/* file to read */
			path = optarg;
			break;
		case 'p':
			/* pipe to read */
			path = optarg;
			break;
		case 'n':
			I->im_path = strdup(optarg);
			break;
		case 's':
			start = strtol(optarg, NULL, 10);
			break;
		case 't':
			if ((I->im_ctx = malloc(sizeof(struct im_file_ctx)))
			    == NULL) {
				dprintf(MSYSLOG_SERIOUS, "om_file_init: error"
				    " allocating context!\n");
				return (-1);
			}
			c = (struct im_file_ctx *) I->im_ctx;
			c->timefmt = strdup(optarg);
			break;
		default:
			dprintf(MSYSLOG_SERIOUS, "om_file_init: command line"
			    " error, at arg %c [%s]\n", ch, optarg? optarg: "");
			return (-1);
		}
	}

	optind = optind_s;	/* restore main's optind */

	if (I->im_ctx != NULL && c->timefmt == NULL) {
		dprintf(MSYSLOG_SERIOUS, "om_file_init: start of time string but"
		    " no string!\n");
		return (-1);
	}

	if (I->im_ctx != NULL)
		c->start = start;

	if (path == NULL) {
		dprintf(MSYSLOG_SERIOUS, "om_file_init: no file/pipe to read\n");
		return (-1);
	}

	if ((I->im_fd = open(path, O_RDONLY, 0)) < 0) {
		dprintf(MSYSLOG_SERIOUS, "im_file_init: can't open %s (%d)\n",
		    argv[1], errno);
		return (-1);
	}

	if (I->im_path == NULL)
		I->im_path = path;	/* no name specified */

	add_fd_input(I->im_fd , I);

	return (1);
}

/*
 * get message
 *
 */

int
im_file_read(struct i_module *im, int infd, struct im_msg  *ret)
{
	char *p, *q;
	int i, c;
	RETSIGTYPE (*sigsave)(int);

        /* ignore sigpipes   for mysql_ping */
        sigsave = place_signal(SIGPIPE, SIG_IGN);

	i = read(im->im_fd, im->im_buf, sizeof(im->im_buf) - 1);
	if (i > 0) {
		(im->im_buf)[i] = '\0';
		for (p = im->im_buf; *p != '\0'; ) {

			/* fsync file after write */
			ret->im_flags = ADDDATE;
			ret->im_pri = DEFSPRI;

			if (im->im_ctx != NULL) {
				struct tm		tm;
				struct im_file_ctx	*c;
				char			*start, *end;

				/* apply strftime */
				c = (struct im_file_ctx *) im->im_ctx;
				if ((end = strptime((p + c->start), c->timefmt, &tm))
				    == NULL) {

					dprintf(MSYSLOG_WARNING, "om_file_read"
					    ": error parsing time!\n");

				} else {

					for (start = p + c->start; *end != '\0';)
						*start++ = *end++;
					*start = '\0';

					if (strftime(ret->im_msg,
					    sizeof(ret->im_msg) - 1,
					    "%b %e %H:%M:%S ", &tm) == 0) {
						dprintf(MSYSLOG_WARNING,
						    "om_file_read: error "
						    "printing time!\n");
					} else {
						ret->im_flags &= !ADDDATE;
					}
				}
			} else if (*p == '<') {
				ret->im_pri = 0;
				while (isdigit((int)*++p))
					ret->im_pri = 10 * ret->im_pri +
					    (*p - '0');
				if (*p == '>')
					++p;
			}

			strncat(ret->im_msg, im->im_path, strlen(ret->im_msg)
			    - sizeof(ret->im_msg) - 1);
			strncat(ret->im_msg, ":", strlen(ret->im_msg)
			    - sizeof(ret->im_msg) - 1);
			dprintf(MSYSLOG_INFORMATIVE, "im_file_read: Entering "
			    "with header %s\n", ret->im_msg);

			if (ret->im_pri &~ (LOG_FACMASK|LOG_PRIMASK))
				ret->im_pri = DEFSPRI;
			q = ret->im_msg + strlen(ret->im_msg);
			while (*p != '\0' && (c = *p++) != '\n' &&
			    q < &ret->im_msg[sizeof(ret->im_msg) - 1]) {
				*q++ = c;
			}
			*q = '\0';
			ret->im_host[0] = '\0';
			ret->im_len = strlen(ret->im_msg);
			logmsg(ret->im_pri, ret->im_msg, ret->im_host,
			    ret->im_flags);
		}
	} else if (i < 0 && errno != EINTR) {
		logerror("im_file_read");
		im->im_fd = -1;
	}

	/* restore previous SIGPIPE handler */
	place_signal(SIGPIPE, sigsave);

	/* if ok return (2) wich means already logged */
	return (im->im_fd == -1 ? -1: 2);
}

int
im_file_close( struct i_module *im)
{
	if (im->im_ctx != NULL) {
		struct im_file_ctx	*c;

		c = (struct im_file_ctx *) im->im_ctx;

		if (c->timefmt)
			free(c->timefmt);
		free(c);
	}

	if (im->im_path != NULL)
		free(im->im_path);

	if (im->im_fd >= 0)
		close(im->im_fd);

	return (0);
}
